File: src/Rmi.js
import { Command } from 'clipanion';

import Base from './Base';

export default class Rmi extends Base {
  static paths = [['rmi']];

  static usage = Command.Usage({
    description: 'Remove the none-images',
    details: `
      This command would find the ids of the none-images and remove them.
    `,
    examples: [[
      'Remove the none images',
      'ddocker rmi',
    ]],
  });

  execute = async () => {
    const ids = (await this.execResult('docker', 'images', '-qf', 'dangling=true'))
      .stdout.toString()
      .split('\n')
      .filter(Boolean);

    if (ids.length === 0) {
      const { stdout } = this.context;

      stdout.write('No none-images need to be removed.\n');
      return;
    }

    await spawn.sync('docker', 'rmi', ...ids);
  };
}

File: src/rmi.rs
use clap::Command;

use crate::utils::{Error, sub_process};

pub fn command() -> Command<'static> {
    Command::new("rmi")
        .about("Find the all ids of the none-images and remove them")
}

pub fn execute() -> Result<(), Error> {
    let stdout = sub_process::exec_result(
        "docker",
        vec!["images", "-qf", "dangling=true"],
    )?;
    let ids: Vec<&str> = stdout.split("\n")
        .filter(|x| !x.is_empty())
        .collect();

    if ids.len() == 0 {
        println!("No none-images need to be removed.");
        return Ok(());
    }

    sub_process::exec(
        "docker",
        [vec!["rmi"], ids]
            .concat(),
    )?;
    Ok(())
}
